cmd/segment-reaper/observer.go

```diff
@@ -143,11 +170,120 @@ func (obsvr *observer) processSegment(ctx context.Context, path metainfo.ScopedP
 // analyzeProject analyzes the objects in obsv.objects field for detecting bad
 // segments and writing them to objs.writer.
 func (obsvr *observer) analyzeProject(ctx context.Context) error {
-	// TODO this part will be implemented in next PR
-	// TODO(if): For what is this?
+	for bucket, objects := range obsvr.objects {
+		for path, object := range objects {
+			if object.skip {
+				continue
+			}
+
+			zombieSegments, err := obsvr.findZombieSegments(object)
+			if err != nil {
+				return err
+			}
+
+			for _, segmentIndex := range zombieSegments {
+				err = obsvr.printSegment(ctx, segmentIndex, bucket, path)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (obsvr *observer) findZombieSegments(object *object) ([]int, error) {
+	zombieSegments := make([]int, 0)
```

Comment:
We could avoid the garbage produced by this if we initialize the slice with a capacity of 64 as a member of the `observer` type. Then pass it as param to this method, which will reset it to 0s as the first thing to do.

Reply:
Actually, no need to reset to 0s. We can just reset the slice length to 0 with `zombieSegments = zombieSegments[:0]`. This will keep the capacity the same, so the slice members won't be garbage collected and can be reused.

Reply:
I was thinking about this but I got no issues with memory during processing 41GB DB but let's do this :)

Reply:
If you don't see any performance issue then it is an optional improvement.

Reply:
changed ;)

Reply:
You could also pass the `zombieBuffer` to `allSegments` to avoid the garbage produced there.